cli/azd/pkg/input/console.go

```diff
@@ -11,17 +11,39 @@ import (
 	"io"
 	"log"
 	"strings"
+	"time"
 
 	"github.com/AlecAivazis/survey/v2"
 	"github.com/azure/azure-dev/cli/azd/pkg/contracts"
 	input_contracts "github.com/azure/azure-dev/cli/azd/pkg/input/contracts"
 	"github.com/azure/azure-dev/cli/azd/pkg/output"
 	"github.com/mattn/go-colorable"
+	"github.com/theckman/yacspin"
+)
+
+type MessageUxType int
+
+const (
+	Title MessageUxType = iota
+	Step
+	StepDone
+	StepFailed
+	ResultSuccess
+	ResultError
 )
 
 type Console interface {
 	// Prints out a message to the underlying console write
 	Message(ctx context.Context, message string)
+	// Prints out a message following the UX format type
+	MessageUx(ctx context.Context, message string, format MessageUxType)
+	// Prints progress spinner with the given title.
+	// If a previous spinner is running, the title is updated.
+	ShowSpinner(ctx context.Context, title string, format MessageUxType)
+	// Stop the current spinner from the console and change the spinner bar for the lastMessage
+	// Set lastMessage to empty string to clear the spinner message instead of a displaying a last message
+	// If there is no spinner running, this is a no-op function
+	StopSpinner(ctx context.Context, lastMessage string, format MessageUxType)
```

Comment:
I would love to see us get to a more composable architecture instead of just adding various new methods onto the `Console` interface.  Right now our cmd/action layer is just too sporadic and we should take our time to introduce a new UX paradigm  that flows nicely from our frontend cmd/action layers down through our backend components.

Reply:
the `console` is already flowing from top to bottom.  And the console is the place where the UX is defined/designed.
I thought about this for a while, and I really think that we should avoid bringing more and more components.

The only goal of UX is adding some specific text formatting within the console-logs. That's why I think having the console supporting this functionality as an extended feature is reasonable.

Note that, the extended methods I am adding to the console is using the underlying capabilities from the console for printing (The Message() method). This means that the logic for the formatters is there.

The new methods I am adding are meant to be just like a middleware between any current component using the console, and the UX standard formatting

Reply:
I'm OK to admit that we have made mistakes in the past when designing some of our components, commands & actions.

This is a great time to correct some of those mistakes.  I currently look at this similar to the following when contrasted to a typically HTTP service.

**cmd** => **HTTP router** (handles how requests are routed to particular controllers
**actions** => Equivalent to the **controller within an MVC pattern**. Primary responsibility to orchestrate the calls to underlying components
**UX** => How the data is displayed to the user (**JSON/xml formatting** for HTTP service) vs Pretty formatted messages with colors, etc in console.

We've made mistakes in the past of having any and all our layers print out messages directly to the console whether it was using our `input.Console` interface or directly via `fmt.Print` commands.

We **NEED** to get to a point where are underlying components are move event based, less tightly coupled to printing and more data contract oriented.

Maybe we need to think of introducing a new `output.Formatter` implementation of some UX/console formatter that can compose a raw stream of events + configuration to achieve the desired output.

Will it take much longer, **Yes**, but it gets us into a better habit of developing loosely coupled components that can scale independently.

Reply:
Here's an example use case to consider:
When output is specified as JSON, we want to move towards emitting typed JSON event messages (based on message type) into `stderr` (what would traditionally be outputted as raw messages into the console. You can see such implementation in `Message`.

"MessageUx" and also "long-running operations" should support structured event messages into `stderr`.

Reply:
`MessageUx` currently works on top of `Message`. The output will behave the same when `--output json` is set.

`MessageUx` is a temporally (WIP) method.  Once that all commands are migrated to use UX output, we can rename `MessageUx -> Message`.  We would just ignore decorating the message with UX format when `--output XXX` is set